### ISSUE TITLE
release fails in jenkins when checked out using a detatched head

### ISSUE THREAD

**Issue by jgangemi** [shillner/unleash-maven-plugin#47]
releases are failing in jenkins and i believe the cause is related to jenkins checking things out in a detached state. i currently set the option to check out to a specific local branch (this is also required by the `maven-release-plugin` to work properly) but the plugin does not seem to notice this.

from the logs:

```
[DEBUG] 14:00:01,078 Git - Push info:
- WORKING_DIR: /var/lib/jenkins/jobs/hello-world-jenkins/workspace
- LOCAL_BRANCH: master
- REMOTE_BRANCH: null
- REMOTE: origin
- REMOTE_URL: git@github.com:condevops/hello-world-jenkins.git
- MERGE_STRATEGY: FULL_MERGE
```

this can be reproduced locally be executing the same steps jenkins does on a check out.

i am currently digging into this and will hopefully have a fix soon but there will be no complaints from me if you beat me to it.

**Comment by shillner** [comment]
I think this issue is rather caused by the git provider. I will create another one there and link it to this one.

**Comment by jgangemi** [comment]
yeah - you're probably right. my bad, i'll try and be a bit more aware of that when filing bugs in the future.

**Comment by umeetiusbaar** [comment]
Is this bug the cause of an error I get in Jenkins:

```
[ERROR] Failed to execute goal com.itemis.maven.plugins:unleash-maven-plugin:2.1.1:perform (default-cli) on project myproject: Execution default-cli of goal com.itemis.maven.plugins:unleash-maven-plugin:2.1.1:perform failed: Could not fetch changes from Git remote 'origin [git@bitbucket.org:myprojects/myproject.git]'. Nothing to fetch. -> [Help 1]
```

with some stacktrace:

```
Caused by: com.itemis.maven.plugins.unleash.scm.ScmException: Could not fetch changes from Git remote 'origin [git@bitbucket.org:myprojects/myproject.git]'.
    at com.itemis.maven.plugins.unleash.scm.providers.ScmProviderGit.update(ScmProviderGit.java:445)
    at com.itemis.maven.plugins.unleash.scm.providers.ScmProviderGit.push(ScmProviderGit.java:373)
    at com.itemis.maven.plugins.unleash.scm.providers.ScmProviderGit.commit(ScmProviderGit.java:337)
    at com.itemis.maven.plugins.unleash.util.DevVersionUtil.commitChanges(DevVersionUtil.java:100)
    at com.itemis.maven.plugins.unleash.steps.actions.SetNextDevVersion.execute(SetNextDevVersion.java:84)
    at com.itemis.maven.plugins.cdi.internal.util.workflow.WorkflowExecutor.executeSequentialWorkflowStep(WorkflowExecutor.java:131)
    at com.itemis.maven.plugins.cdi.internal.util.workflow.WorkflowExecutor.execute(WorkflowExecutor.java:114)
    at com.itemis.maven.plugins.cdi.AbstractCDIMojo.execute(AbstractCDIMojo.java:208)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
    ... 32 more
```

And similar Push info as @jgangemi reported above:

```
[INFO] 09:28:14,970  Git - Pushing local changes to remote repository.
[DEBUG] 09:28:14,971  Git - Push info:
    - WORKING_DIR: /var/lib/jenkins/workspace/myproject
    - LOCAL_BRANCH: master
    - REMOTE_BRANCH: null
    - REMOTE: origin
    - REMOTE_URL: git@bitbucket.org:myprojects/myproject.git
    - MERGE_STRATEGY: FULL_MERGE

[INFO] 09:28:14,972  Git - Updating local repository with remote changes.
[DEBUG] 09:28:14,973  Git - Fetching remote updates.
[DEBUG] 09:28:14,978  Git - Fetch info:
    - WORKING_DIR: /var/lib/jenkins/workspace/myproject
    - REMOTE: origin
    - REMOTE_URL: git@bitbucket.org:myprojects/myproject.git
[INFO] 09:28:14,980  Rollback of all pom changes necessary for setting of the development version as well as reverting any made SCM commits.
```

It seems to fail at the "Fetching remote updates" phase.

It also gives an error during the rollback phase even though it seems to go through okay in the end (no tags or commits in remote repo):

```
[ERROR] Error calling rollback method of Mojo.
java.lang.reflect.InvocationTargetException
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:498)
    at com.itemis.maven.plugins.cdi.internal.util.workflow.WorkflowExecutor.rollback(WorkflowExecutor.java:227)
    at com.itemis.maven.plugins.cdi.internal.util.workflow.WorkflowExecutor.rollback(WorkflowExecutor.java:206)
    at com.itemis.maven.plugins.cdi.internal.util.workflow.WorkflowExecutor.executeSequentialWorkflowStep(WorkflowExecutor.java:133)
    at com.itemis.maven.plugins.cdi.internal.util.workflow.WorkflowExecutor.execute(WorkflowExecutor.java:114)
    at com.itemis.maven.plugins.cdi.AbstractCDIMojo.execute(AbstractCDIMojo.java:208)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:207)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307)
    at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193)
    at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106)
    at org.jvnet.hudson.maven3.launcher.Maven32Launcher.main(Maven32Launcher.java:132)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchStandard(Launcher.java:330)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:238)
    at jenkins.maven3.agent.Maven32Main.launch(Maven32Main.java:186)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:498)
    at hudson.maven.Maven3Builder.call(Maven3Builder.java:136)
    at hudson.maven.Maven3Builder.call(Maven3Builder.java:71)
    at hudson.remoting.UserRequest.perform(UserRequest.java:153)
    at hudson.remoting.UserRequest.perform(UserRequest.java:50)
    at hudson.remoting.Request$2.run(Request.java:332)
    at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:68)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.IllegalStateException: No from revision set (start revision of the reversion)!
    at com.google.common.base.Preconditions.checkState(Preconditions.java:174)
    at com.itemis.maven.plugins.unleash.scm.requests.RevertCommitsRequest$Builder.build(RevertCommitsRequest.java:187)
    at com.itemis.maven.plugins.unleash.steps.actions.SetNextDevVersion.rollback(SetNextDevVersion.java:129)
    ... 42 more
```

**Comment by jgangemi** [comment]
i believe so - that's the same error i was getting when i tried to do the release via jenkins. 

unfortunately i haven't been able to put in any cycles on this yet.

**Comment by shillner** [comment]
I will schedule this detached HEAD thing for this week as it affects us too!

**Comment by shillner** [comment]
Fixed in git provider version 2.0.1
